src/kernels/jupyter/finder/universalRemoteKernelFinder.ts

```diff
@@ -0,0 +1,397 @@
+        const ui = new DisplayOptions(false);
+        return this.notebookProvider.connect({
+            resource: undefined,
+            ui,
+            localJupyter: false,
+            token: cancelToken,
+            serverId: this.serverUri.serverId
+        });
+    }
+
+    private async getFromCache(cancelToken?: CancellationToken): Promise<RemoteKernelConnectionMetadata[]> {
+        let results: RemoteKernelConnectionMetadata[] = this.cache;
+        const key = this.getCacheKey();
+
+        // If not in memory, check memento
+        if (!results || results.length === 0) {
+            // Check memento too
+            const values = this.globalState.get<{
+                kernels: RemoteKernelConnectionMetadata[];
+                extensionVersion: string;
+            }>(key, { kernels: [], extensionVersion: '' });
+
+            if (values && isArray(values.kernels) && values.extensionVersion === this.env.extensionVersion) {
+                results = values.kernels.map(deserializeKernelConnection) as RemoteKernelConnectionMetadata[];
+                this.cache = results;
+            }
+        }
+
+        // Validate
+        const validValues: RemoteKernelConnectionMetadata[] = [];
+        const promise = Promise.all(
+            results.map(async (item) => {
+                if (await this.isValidCachedKernel(item)) {
+                    validValues.push(item);
+                }
+            })
+        );
+        if (cancelToken) {
+            await Promise.race([
+                promise,
+                createPromiseFromCancellation({ token: cancelToken, cancelAction: 'resolve', defaultValue: undefined })
+            ]);
+        } else {
+            await promise;
+        }
+        return validValues;
+    }
+
+    // Talk to the remote server to determine sessions
+    @capturePerfTelemetry(Telemetry.KernelListingPerf, { kind: 'remote' })
+    public async listKernelsFromConnection(
+        connInfo: INotebookProviderConnection
+    ): Promise<RemoteKernelConnectionMetadata[]> {
+        // Get a jupyter session manager to talk to
+        let sessionManager: IJupyterSessionManager | undefined;
+        // This should only be used when doing remote.
+        if (connInfo.type === 'jupyter') {
+            try {
+                sessionManager = await this.jupyterSessionManagerFactory.create(connInfo);
+
+                // Get running and specs at the same time
+                const [running, specs, sessions, serverId] = await Promise.all([
+                    sessionManager.getRunningKernels(),
+                    sessionManager.getKernelSpecs(),
+                    sessionManager.getRunningSessions(),
+                    computeServerId(connInfo.url)
+                ]);
+
+                // Turn them both into a combined list
+                const mappedSpecs = await Promise.all(
+                    specs.map(async (s) => {
+                        await sendKernelSpecTelemetry(s, 'remote');
```

Comment:
i think we don't need the await here, it does the whole process of fetching the kernels